Data/ByteString/Base64/Internal.hs

```diff
@@ -160,7 +160,7 @@ decodeWithTable padding !decodeFP bs
     noPad = "Base64-encoded bytestring required to be unpadded"
     invalidPad = "Base64-encoded bytestring has invalid padding"
 
-    !dlen = q * 3
+    !dlen = (q + signum r) * 3
```

Comment:
`B.length bs` is non-negative, so `r` is non-negative as well. Why `signum`?

Reply:
I'd suggest ``!dlen = (B.length bs + 3) `quot` 4`` instead, giving GHC an opportunity to avoid slow division.

Reply:
I think sugnum for the case where r is 0. In other words, only add the factor when theres a remainder (taking the ceiling)

Reply:
Tbh @Bodigrim is right here and I agree with him. I think the solution to this is to move the length calculation down into the go block as opposed to calculating it here which is too early. That should fix the problems with malloc and I'm sure we use correct length

Reply:
@Bodigrim can you please explain the performance aspects of `quot` vs `div`?  I was not aware.

Also, your suggested implementation is missing the `(*3)`.  So it should be `!dlen = ((B.length bs + 3) `quot` 4) * 3`.

Regarding `signum`, @brandonchinn178's explanation is correct.

Reply:
@frasertweedale `quotRem` rounds quotient towards 0 (so ``(-1) `quotRem` 4 = (0,-1)``), while `divMod` rounds quotient towards negative infinity (so ``(-1) `divMod` 4 = (-1, 3)``). Usually `quotRem` is a more primitive operation and on the majority of platforms compiles immediately to a CPU instruction, while `divMod` requires several conditional branches. See definitions here: https://hackage.haskell.org/package/base-4.14.0.0/docs/src/GHC.Base.html#quotRemInt

Since for positive inputs `quotRem` and `divInt` match, one should prefer `quotRem` in our case.

Now, why I'm unhappy with `signum`? Exactly for the same reason: it brings conditional branches into a code, which can otherwise be branchless. 

Finally, division (even `quotRem` is a very slow operation). Newer GHCs have a special optimization for `quot` (but not for `quotRem`), which replaces division by a power of two with a bit shift. So if you write `!dlen = ((B.length bs + 3) `quot` 4) * 3`, it will compile to a couple of simple CPU instructions. It would be even better to use ``(`shiftL` 2)`` instead of ``(`quot` 4)`` to be explicit. 

This is tangential to the issue at hand, thanks for catching and fixing it indeed!

Reply:
@Bodigrim many thanks for the detailed explanation.